### ISSUE TITLE
[Feature Request]: Form Groups

### ISSUE THREAD

**Issue by crutchcorn** [TanStack/form#419]
### Description

When building a form stepper, like so:

![image](https://user-images.githubusercontent.com/9100169/228072732-7514d64c-b6f5-471e-9b0b-e598597202a7.png)

It's common for each step to have its own form. However, this complicates the form submission and validation process by requiring you to add complex logic.

Ideally, it would be nice to have a FormGroup where you could validate the group, but not the form itself - `submit` the value and move on to the next step.

### Describe the solution you'd like

What if there was something like this psuedo-API?

```tsx
import { useState } from "react";
import { useForm } from "@tanstack/react-form";
// Import might change
import { za } from "@tanstack/zod-form-adapter";

const StepperForm = () => {
  const [step, setStep] = useState(0);
  const form = useForm({ onSubmit: () => {} });

  return (
    <form.Provider>
      <form {...form.getFormProps()}>
        <div>
          {step === 1 && (
            <form.FormGroup
              onGroupSubmit={() => {
                setStep(step + 1);
              }}
              preserveState={true}
            >
              <form.Field
                name="step1.name"
                onSubmit={za(z.string().thing())}
                onGroupSubmit={za(z.string().thing())}
              >
                <Input />
              </form.Field>
            </form.FormGroup>
          )}

          {step === 2 && (
            <form.FormGroup onGroupSubmit={submit} preserveState={true}>
              <form.Field
                name="step2.name"
                onSubmit={za(z.string().thing())}
                onGroupSubmit={za(z.string().thing())}
              >
                <Input />
              </form.Field>
            </form.FormGroup>
          )}
        </div>
      </form>
    </form.Provider>
  );
};
```

**Comment by Christian24** [comment]
Just wanted to leave some additional context: We sometimes have quite complex forms that have a lot of fields that are only visible on a certain condition, say for example a billing and a delivery address, they are basically the same, but one is only used under the condition that the two are not the same. So FormGroups can be a great way to structure a form as well.

**Comment by timothyac** [comment]
+1

At AWS, a common pattern we implement is a “Wizard” pattern which guides a user through a multi-step form. 

Example: https://cloudscape.design/examples/react/wizard.html

**Comment by francesconi** [comment]
I'm currently working on a multi-step form, similar to a wizard. The proposed `form.FormGroup` component would simplify the code a lot.

**Comment by crutchcorn** [comment]
@timothyac we're not quite ready for this feature yet as part of our 1.x release, but we'll work on it soon after!

That said, I'd love to help the AWS team figure out how to integrate Cloudscape with TanStack Form when it's ready. If y'all need any help with anything, let me know (even via DMs - they're open)

**Comment by samuellawerentz** [comment]
Any update on the feature?

**Comment by vishal-tiwari** [comment]
Any update on this feature?

**Comment by paulm17** [comment]
Need this too, get back to work!

![Image](https://github.com/user-attachments/assets/dd4fc39d-e015-4fb6-be7a-ee9af00cf42c)

**Comment by ataylorme** [comment]
@paulm17 what a terrible comment to leave, even if it is in jest.

**Comment by paulm17** [comment]
from the discord: https://discord.com/channels/719702312431386674/1100437019857014895/1347537741704990720

![Image](https://github.com/user-attachments/assets/56c40e03-c7bf-477c-ae76-f387407d257e)

Don't be so sensitive.

**Comment by rudzky** [comment]
Any update on this feature? Have anybody found some solution for multi-step forms? 😁

**Comment by ataylorme** [comment]
I found this but haven’t tested it yet https://www.formity.app/

**Comment by danhaus** [comment]
There's a PR: https://github.com/TanStack/form/pull/1469

**Comment by remyoudemans** [comment]
> There's a PR: [#1469](https://github.com/TanStack/form/pull/1469)

Maybe I've misunderstood this issue or what that PR added, but they seem quite different to me. This issue is looking for a way to handle submission (and validation) of a form group separately from the whole form so we could easily build multi-step forms.
I don't think `withFieldGroup` really solves this problem.

**Comment by FatahChan** [comment]
Hello, want to pitch in an idea here, instead of creating a group for the subforms (each step), we could make each step a form, and have a group with  to tie them all together, here is how would look like.





```tsx
const Form1 = () =>{

  const form = useForm({ onSubmit: () => {
     // got to nextstep
   } });
  return (
     /// usual form
   }
}


const Form2 = () =>{

  const form = useForm({ onSubmit: () => {
         // got to nextstep
  } });
  return (
     /// usual form
   }
}

const Form3 = () =>{

 const formGroup = useFormGroupContext()

  const form = useForm({ onSubmit: () => {
      formGroup.submit() 
   } });
  return (
     /// usual form
   }
}

const MultistepForm = () => {

const formGroup = useFormGroup({onSubmit: (values) => {
  console.log(values)
}})
 return (
    <formGroup.Provider/>
    
    {step == 1 & <Form1/>}
    {step == 2 & <Form2/>}
    {step == 3 & <Form3/>}
    
    <formGroup.Provider/>
   )
}

```

just throwing an idea out here, but I guess typing can't be inferred at the useFormGroup, so so manual typing is needed.

**Comment by Frost7994** [comment]
Any update on how to implement something like this?

**Comment by jmn-lime** [comment]
What's the current recommended approach to handle this at the moment?

**Comment by adelinerohr** [comment]
I am looking to build a complex onboarding wizard right now and would love some guidance!

**Comment by 413n** [comment]
At the moment as a workaround we've used withFieldGroup on each step form to have their own schema and being able to map them to another key on the parent. Then we have created a custom onSubmit validation that based on the active step it validates only the schema coming from it. To "submit" each step (except the last) we used workaround so that we trigger only the validation `const result = form.validateSync("submit")` and based on the result we go to the next one. In the last step we both run the validation and also do `form.handleSubmit()`.
It's not the best but wrapped in a custom hook it's "good enough"

**Comment by zhouzi** [comment]
I think the main missing pieces in the current API are:

1. A way to submit a group. The challenge is that relying on a native submit event would imply nesting `<form>` elements, which isn't possible. One potential approach could be to let a group inject some form of "middleware" into the submission flow.
2. A way to read a group's state and errors. This one seems more straightforward, since most of the underlying mechanisms already exist.

**Comment by crutchcorn** [comment]
Hey all! Just wanted to leave an update here that we _are_ actively working on this API.

To that end @LeCarbonator, @harry-whorlow, and myself have brainstormed the following API that we've updated in the main API:

https://github.com/TanStack/form/issues/419#issue-1873832227

We'll begin implementations here soon, but wanted to open the discussion up publicly more to gain any feedback folks might have.

**Comment by alexbaileyuk** [comment]
> Hey all! Just wanted to leave an update here that we _are_ actively working on this API.
> 
> To that end [@LeCarbonator](https://github.com/LeCarbonator), [@harry-whorlow](https://github.com/harry-whorlow), and myself have brainstormed the following API that we've updated in the main API:
> 
> [#419 (comment)](https://github.com/TanStack/form/issues/419#issue-1873832227)
> 
> We'll begin implementations here soon, but wanted to open the discussion up publicly more to gain any feedback folks might have.

It's hard to know for sure until we can get hands on with it but I'm desperate for an API which works the way you've described.

Using that suggested approach, would we be able to subscribe to individual form groups for custom components? As an example, I'd like to have a "Continue" button on each step, but it should be disabled if the form group is not valid rather than the user trying to click it and then finding out it's invalid.

**Comment by iivvaannxx** [comment]
While waiting for an official implementation I've managed to put something on my end that I find quite comfortable to work with (for my use case). See below an example on how do I declare my multi step form (Vue composable). 

It's fully type-safe from end to end and it uses one single form.

```ts
// use-registration-form.ts
import { z } from "zod";

// 1. Define the Schema
const registrationSchema = z.object({
  firstName: z.string().min(1),
  lastName: z.string().min(1),
  email: z.string().email(),
  password: z.string().min(8),
  newsletter: z.boolean(),
});

type RegistrationSchema = typeof registrationSchema;

// 2. Define the Steps
// Note: the `fields` in each step are automatically validated against the schema before changing step
const STEPS = [
  {
    name: "personalInfo",
    fields: ["firstName", "lastName"],
  },
  {
    name: "accountDetails",
    fields: ["email", "password"],
  },
  {
    name: "preferences",
    fields: ["newsletter"],

    // Example: Only show this step if the user provided a specific email domain
    include: (state) => !state.values.email.endsWith("@internal.com"),
  },
] as const satisfies MultiStepTanstackFormStep<RegistrationSchema>[];

// 3. Define Context (Optional)
// Custom data you may want to have available at any given moment for some conditional logic
type RegistrationContext = {
  /**
   * The role of the person filling out the form.
   * Admins might see extra fields (like "Internal Notes" or "Role Assignment").
   */
  userRole: "admin" | "guest";

  /**
   * Feature flags to toggle specific form sections.
   */
  features?: {
    enableNewsletter?: boolean;
  };
};

type UseRegistrationFormParams = {
  context: RegistrationContext;
  onSubmit: (values: z.output<RegistrationSchema>) => void;
};

// 4. Create the Composable
function __useRegistrationForm(params: UseRegistrationFormParams) {
  return useMultiStepTanstackForm({
    ...params,
    schema: registrationSchema,
    steps: STEPS,
    defaultValues: {
      firstName: "",
      lastName: "",
      email: "",
      password: "",
      newsletter: false,
    },
  });
}

// 5. Create the Form Kit
// This generates the provider, injection key, and typed hooks
export const registrationFormKit = createMultiStepTanstackForm(
  "registration", // Unique name for the form
  STEPS,
  __useRegistrationForm
);

// 6. Export Reusable Utilities
// Composables to create a new form instance, inject an existing one or an step
export const useRegistrationForm = registrationFormKit.useForm;
export const injectRegistrationForm = registrationFormKit.injectForm;
export const useRegistrationFormStep = registrationFormKit.useStep;

// A provider component (to use instead of "useRegistrationForm") and a the root Form component.
export const RegistrationFormProvider = registrationFormKit.ProviderComponent;
export const RegistrationForm = registrationFormKit.FormComponent;

// Types
type FormHook = typeof useRegistrationForm;
export type RegistrationFormData = InferTanstackFormData<FormHook>;
export type RegistrationFormApi = InferTanstackFormApi<FormHook>;
export type RegistrationFormField = InferTanstackFormFieldApi<FormHook>;
```

You then use it like below. Someone mentioned having a component per-step is inconvenient but I prefer it that way to keep everything organized and cleaner.

<details>

<summary><b>registration-form.vue</b></summary>

```vue
<script setup lang="ts">
// registration-form.vue

// Handle the valid form submission
async function handleSubmit(values: TypedSchemaOutput) {
  console.log("Form Submitted!", values);
  // await api.registerUser(values);
}
</script>

<template>
    <!-- 1. Wrap everything in the Provider (or use composable in setup script)  -->
    <!-- Pass the context required by your composable -->
    <RegistrationFormProvider
      :context="{ mode: 'create', userRole: 'guest' }"
      @submit:values="handleSubmit"
    >
      <!-- The default slot exposes all the data and validation utilities -->
      <template #default="{ stepper }">
        <!-- 2. Use the Form Component wrapper -->
        <RegistrationForm>
            <!-- 3. Render the Wizard Steps (isCurrent type-safely infers the step names as a string union) -->
            <StepPersonalInfo v-if="stepper.isCurrent('personalInfo')" />
            <StepAccountDetails v-if="stepper.isCurrent('accountDetails')" />
            <StepPreferences v-if="stepper.isCurrent('preferences')" />
        </RegistrationForm>
      </template>
    </RegistrationFormProvider>
</template>
```
</details>


<details>

<summary><b>step-personal-info.vue</b></summary>

```vue
<script setup lang="ts">
// step-personal-info.vue

import { 
  injectRegistrationForm, 
  useRegistrationFormStep 
} from "../use-registration-form";

// 1. Inject the main form API (for fields, state, etc.)
const { Form } = injectRegistrationForm();

// 2. Use the step hook to get typed metadata for THIS specific step
// This ensures 'step' is strictly typed as the 'personalInfo' step definition
const step = useRegistrationFormStep("personalInfo");

// Example: Accessing custom properties defined in the step definition
// console.log(step.name); // "personalInfo"
</script>

<template>
  <div>
    <Form.Field name="firstName">
       <!-- Your field template -->
    </Form.Field>
  
    <Form.Field name="lastName">
      <!-- Your field template -->
    </Form.Field>
  <div>
</template>
```

</details>


I don't know if the official implementation will take too much time or not, in the meantime, if any of you are interested on how this works let me know and I'll try to find some time to write about it

**Comment by devnaumov** [comment]
Why can’t we just use the withForm api?
We can wrap each step in withForm and still keep a single form
https://tanstack.com/form/v1/docs/framework/solid/guides/form-composition

**Comment by alexbaileyuk** [comment]
> Why can’t we just use the withForm api? We can wrap each step in withForm and still keep a single form https://tanstack.com/form/v1/docs/framework/solid/guides/form-composition

I gave this a go for a few days on a React Native app recently and every approach I took resulted in thousands of lines of sprawling mess that was extremely hard to follow. I'm not sure I'm fully clear on how to wrap each step in a different `withForm` and still keep a single form?

**Comment by zhouzi** [comment]
> Why can’t we just use the withForm api?
> We can wrap each step in withForm and still keep a single form
> https://tanstack.com/form/v1/docs/framework/solid/guides/form-composition

You can totally do it. But you can't validate each step individually before moving to the next step ("natively", at least).

**Comment by crutchcorn** [comment]
@joshua-singer please keep comments productive and forward-looking. We know it's a pain-point, but without feedback on the RFC we're proposing comments like yours only serve to add frustration for us maintainers.

Let us know if the RFC proposed in OP meets your needs or not.

**Comment by joshua-singer** [comment]
Sorry man, never thought about that @crutchcorn 🫶 the proposal looks good!

**Comment by mikecebul** [comment]
I think I had this all wrong. I thought I could use field groups for multi steps and I think it only kinda worked because I lost some type safety. I am refactoring with useForm for steps rn.

The big issue this future api solves for me is being able to have field level validators. Rn I need to keep all validation at the form level or else an onchange field validation will freeze my form from going forward, if I have a field error and go back a step.

I just setup nuqs as the source of truth for step navigation and I like it. Will I be able to share a type between nuqs and the Form Group?

**Comment by JBezerra** [comment]
We've solved this by treating each step as a completely independent form rather than using nested form groups. Here's our approach:

**Separate forms per step**: Each step is its own form component with its own `useAppForm` hook and schema. No shared form state between steps.

**Step navigation via React's Context API**: We use a `StepFormContext` to manage step state (`currentStep`, `goToNextStep`, `goToPrevStep`). This keeps navigation logic separate from form logic.

**Independent validation**: Each form validates only its own fields. When you submit a step, it validates that step's schema, persists the data via API, then calls `goToNextStep()` on success.

**Example structure**:
```tsx
// StepFormContext manages navigation
const { goToNextStep } = useStepForm()

// Each step is an independent form
const form = useAppForm({
  defaultValues: { /* step-specific defaults */ },
  validators: { onSubmit: stepSchema },
  onSubmit: async ({ value }) => {
    await mutateStepData(value) // API call
    goToNextStep() // Navigate on success
  },
})
```

This avoids the complexity of nested form contexts and conditional validation. Each step is self-contained, which makes the code easier to reason about and maintain. The trade-off is that you can't validate the entire form at once, but for steppers where you want to persist data incrementally, this works well.

**Comment by onyedikachi23** [comment]
@JBezerra This is how I've always done it, but using React Hook Form. 

The proposal makes sense, but I don't see how it's better than having those steps as independent forms.

You can always cook up some custom logic when you need to chain some logic across the steps. With Claude Code, this won't be pain the ass.

**Comment by hanzlah101** [comment]
@crutchcorn any update?

**Comment by cgatian** [comment]
For people looking for a pretty nice solution to this, checkout https://github.com/martiserra99/formity

**Comment by aladicf** [comment]
> @crutchcorn any update?

They have decided and settled on the API, Corbin is starting on the implementation this month, hopefully us devs will have something to test out at the end of the month or early next month. 🫡

**Comment by crutchcorn** [comment]
Ground has officially broke on development!:

https://github.com/TanStack/form/pull/2128

We have our first few core tests passing and I've passed the rough sketch of our initial impl to the maintaining team for review.

I hope to move fast, but given the length of the TODO, it might be a moment until this lands in stable. I'll be sure to post major updates here, especially when we have a version of the React adapter ready for you lovely folk to test.

Thanks again for sticking with us! 🚀

**Comment by monarcode** [comment]
Just checked the PR again, patiently waiting for this to ship

**Comment by crutchcorn** [comment]
I believe that we're ready for some initial public testers of #2128. All features that will eventually land in production should be supported, but bugs may be present and docs are not yet written. Likewise, the only adapter that's currently supported is React.

If you're interested in testing, you can do so here:

```shell
npm i https://pkg.pr.new/@tanstack/react-form@2128
```

And refer to the `examples/react/multi-step-wizard` example for usage (temporary until docs are added).

If you find any issues, please report them as comments in the PR.

I'll continue to update this thread when docs or other major adapters land.

Thanks!

**Comment by crutchcorn** [comment]
Quick update for you all:

We now have FormGroup support in every framework adapter (including our brand new Preact adapter).

https://github.com/TanStack/form/pull/2128

We're still missing docs (coming soon), but this has been a fairly big lift to get where it is now. In fact, to make sure the experience is as good as possible regardless of what framework you're coming from, we:

- [Found and fixed an issue that allowed `field.Field` needlessly](https://github.com/TanStack/form/pull/2149)
- [Added a `withForm`-like API for Angular](https://github.com/TanStack/form/releases/tag/%40tanstack%2Fangular-form%401.31.0)
- [Fixed a typing issue with Vue's `form.Subscribe` component](https://github.com/TanStack/form/releases/tag/%40tanstack%2Fvue-form%401.29.3)
- [Added a type helper to easily get the `form` type into another component for Svelte](https://github.com/TanStack/form/releases/tag/%40tanstack%2Fsvelte-form%401.30.0)
- [Added that same type helper to Lit](https://github.com/TanStack/form/releases/tag/%40tanstack%2Flit-form%401.24.0)
- [Ported TanStack Store to Lit](https://github.com/TanStack/store/releases/tag/%40tanstack%2Flit-store%400.13.2)
- [Fixed TanStack Store's Preact adapter to handle unstable selector functions now](https://github.com/TanStack/store/releases/tag/%40tanstack%2Fpreact-store%400.13.1)

I'm excited to get everyone's hands on this new feature. I'll update this PR when we have docs. In the meantime, if you'd like to test the code in your framework of choice, see this PR comment from `pkg-pr-new` bot: https://github.com/TanStack/form/pull/2128#issuecomment-4258714258

**Comment by mikecebul** [comment]
Hey Corbin,

Hope you're enjoying your Memorial Day Weekend! Thanks for this release. I got 6 multi step forms switched over to the new api with the react adapter. Everything seems to be working properly and ready to merge with a pr. Unless you are expecting breaking changes before you merge then I'll wait, but I'm ready to give it a shot and am already working on my next big PR that depends on these changes. 
https://github.com/mikecebul/drug-test-mi/pull/22/changes

**Comment by crutchcorn** [comment]
@mikecebul thank you for continually being an awesome tester and community member! 🎉

No breaking changes from that PR are planned and we should be merging shortly. I got swept into our Router security incident, which set me back by a bit, but I'm back in the saddle now.

**Comment by crutchcorn** [comment]
My implementation of Form Groups have been completed, docs and examples included:

https://github.com/TanStack/form/pull/2128

I'm waiting on a review from the other maintainers, but that should be a quick process, since they reviewed things early on when the code was less stable. We'll likely publish this in the next few days.

While y'all are waiting for a release, please go ahead and test the code in your projects like @mikecebul did. It will help us find any potential errors before going live.

Thanks!

**Comment by crutchcorn** [comment]
We're live!

Closing as the feature is now shipped.

https://tanstack.com/form/latest/docs/framework/react/guides/form-groups
